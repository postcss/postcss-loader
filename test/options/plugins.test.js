import {
  compile,
  getCompiler,
  getErrors,
  getCodeFromBundle,
  getWarnings,
} from '../helpers/index';

describe('Options Plugins', () => {
  it('should work Plugins - {Array}', async () => {
    const compiler = getCompiler('./css/index.js', {
      ident: 'postcss',
      // eslint-disable-next-line global-require
      plugins: [require('../fixtures/config/plugin')()],
    });
    const stats = await compile(compiler);

    const codeFromBundle = getCodeFromBundle('style.css', stats);

    expect(codeFromBundle.css).toMatchSnapshot('css');
    expect(getWarnings(stats)).toMatchSnapshot('warnings');
    expect(getErrors(stats)).toMatchSnapshot('errors');
  });

  it('should work Plugins - {Object}', async () => {
    const compiler = getCompiler('./css/index.js', {
      ident: 'postcss',
      // eslint-disable-next-line global-require
      plugins: require('../fixtures/config/plugin'),
    });
    const stats = await compile(compiler);

    const codeFromBundle = getCodeFromBundle('style.css', stats);

    expect(codeFromBundle.css).toMatchSnapshot('css');
    expect(getWarnings(stats)).toMatchSnapshot('warnings');
    expect(getErrors(stats)).toMatchSnapshot('errors');
  });

  it('should work Plugins - {Function} - {Array}', async () => {
    const compiler = getCompiler('./css/index.js', {
      ident: 'postcss',
      // eslint-disable-next-line global-require
      plugins: () => [require('../fixtures/config/plugin')()],
    });
    const stats = await compile(compiler);

    const codeFromBundle = getCodeFromBundle('style.css', stats);

    expect(codeFromBundle.css).toMatchSnapshot('css');
    expect(getWarnings(stats)).toMatchSnapshot('warnings');
    expect(getErrors(stats)).toMatchSnapshot('errors');
  });

  it('should work Plugins - {Function} - {Object}', async () => {
    const compiler = getCompiler('./css/index.js', {
      ident: 'postcss',
      // eslint-disable-next-line global-require
      plugins: () => require('../fixtures/config/plugin')(),
    });
    const stats = await compile(compiler);

    const codeFromBundle = getCodeFromBundle('style.css', stats);

    expect(codeFromBundle.css).toMatchSnapshot('css');
    expect(getWarnings(stats)).toMatchSnapshot('warnings');
    expect(getErrors(stats)).toMatchSnapshot('errors');
  });

  it('should work Plugins - {Object without require}', async () => {
    const compiler = getCompiler('./css/index.js', {
      plugins: {
        'postcss-import': {},
        'postcss-nested': {},
      },
    });
    const stats = await compile(compiler);

    const codeFromBundle = getCodeFromBundle('style.css', stats);

    expect(codeFromBundle.css).toMatchSnapshot('css');
    expect(getWarnings(stats)).toMatchSnapshot('warnings');
    expect(getErrors(stats)).toMatchSnapshot('errors');
  });

  it('should work Plugins - {empty Object}', async () => {
    const compiler = getCompiler('./css/index.js', {
      plugins: {},
    });
    const stats = await compile(compiler);

    const codeFromBundle = getCodeFromBundle('style.css', stats);

    expect(codeFromBundle.css).toMatchSnapshot('css');
    expect(getWarnings(stats)).toMatchSnapshot('warnings');
    expect(getErrors(stats)).toMatchSnapshot('errors');
  });

  it('should work Plugins - {Object without require} + options', async () => {
    const compiler = getCompiler('./css/index2.js', {
      plugins: {
        'postcss-short': { prefix: 'x' },
      },
      config: false,
    });
    const stats = await compile(compiler);

    const codeFromBundle = getCodeFromBundle('style2.css', stats);

    expect(codeFromBundle.css).toMatchSnapshot('css');
    expect(getWarnings(stats)).toMatchSnapshot('warnings');
    expect(getErrors(stats)).toMatchSnapshot('errors');
  });

  it('should work Plugins - {Object} + options', async () => {
    const compiler = getCompiler('./css/index2.js', {
      // eslint-disable-next-line global-require
      plugins: require('postcss-short')({ prefix: 'x' }),
      config: false,
    });
    const stats = await compile(compiler);

    const codeFromBundle = getCodeFromBundle('style2.css', stats);

    expect(codeFromBundle.css).toMatchSnapshot('css');
    expect(getWarnings(stats)).toMatchSnapshot('warnings');
    expect(getErrors(stats)).toMatchSnapshot('errors');
  });

  it('should work Plugins - {Array<Object>} + options', async () => {
    const compiler = getCompiler('./css/index2.js', {
      // eslint-disable-next-line global-require
      plugins: [require('postcss-short')({ prefix: 'x' })],
      config: false,
    });
    const stats = await compile(compiler);

    const codeFromBundle = getCodeFromBundle('style2.css', stats);

    expect(codeFromBundle.css).toMatchSnapshot('css');
    expect(getWarnings(stats)).toMatchSnapshot('warnings');
    expect(getErrors(stats)).toMatchSnapshot('errors');
  });

  it('should disables plugin from config', async () => {
    const compiler = getCompiler('./css/index2.js', {
      config: 'test/fixtures/css/plugins.config.js',
      plugins: {
        'postcss-short': false,
      },
    });
    const stats = await compile(compiler);

    const codeFromBundle = getCodeFromBundle('style2.css', stats);

    expect(codeFromBundle.css).toMatchSnapshot('css');
    expect(getWarnings(stats)).toMatchSnapshot('warnings');
    expect(getErrors(stats)).toMatchSnapshot('errors');
  });

  it('should emit error on load plugin', async () => {
    const compiler = getCompiler('./css/index2.js', {
      plugins: {
        'postcss-unresolved': {},
      },
    });
    const stats = await compile(compiler);

    expect(getWarnings(stats)).toMatchSnapshot('warnings');
    expect(getErrors(stats, true)).toMatchSnapshot('errors');
  });
});
