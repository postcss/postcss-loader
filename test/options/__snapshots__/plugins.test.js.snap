// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Options Plugins should disables plugin from config: css 1`] = `
"a {
  -x-border-color: blue blue *;
  -x-color: * #fafafa;
}
"
`;

exports[`Options Plugins should disables plugin from config: errors 1`] = `Array []`;

exports[`Options Plugins should disables plugin from config: warnings 1`] = `Array []`;

exports[`Options Plugins should emit error on load plugin: errors 1`] = `
Array [
  "ModuleBuildError: Module build failed (from \`replaced original path\`):
TypeError: Cannot read property 'postcss' of undefined",
  "ModuleError: Module Error (from \`replaced original path\`):
Loading PostCSS Plugin failed: Cannot find module 'postcss-unresolved' from 'src/utils.js'",
]
`;

exports[`Options Plugins should emit error on load plugin: warnings 1`] = `Array []`;

exports[`Options Plugins should work Plugins - {Array<Object>} + options: css 1`] = `
"a {
  border-top-color: blue;
  border-right-color: blue;
  border-left-color: blue;
  background-color: #fafafa;
}
"
`;

exports[`Options Plugins should work Plugins - {Array<Object>} + options: errors 1`] = `Array []`;

exports[`Options Plugins should work Plugins - {Array<Object>} + options: warnings 1`] = `Array []`;

exports[`Options Plugins should work Plugins - {Array}: css 1`] = `
"a { color: rgba(255, 0, 0, 1.0) }
"
`;

exports[`Options Plugins should work Plugins - {Array}: errors 1`] = `Array []`;

exports[`Options Plugins should work Plugins - {Array}: warnings 1`] = `Array []`;

exports[`Options Plugins should work Plugins - {Function} - {Array}: css 1`] = `
"a { color: rgba(255, 0, 0, 1.0) }
"
`;

exports[`Options Plugins should work Plugins - {Function} - {Array}: errors 1`] = `Array []`;

exports[`Options Plugins should work Plugins - {Function} - {Array}: warnings 1`] = `Array []`;

exports[`Options Plugins should work Plugins - {Function} - {Object}: css 1`] = `
"a { color: rgba(255, 0, 0, 1.0) }
"
`;

exports[`Options Plugins should work Plugins - {Function} - {Object}: errors 1`] = `Array []`;

exports[`Options Plugins should work Plugins - {Function} - {Object}: warnings 1`] = `Array []`;

exports[`Options Plugins should work Plugins - {Object without require} + options: css 1`] = `
"a {
  border-top-color: blue;
  border-right-color: blue;
  border-left-color: blue;
  background-color: #fafafa;
}
"
`;

exports[`Options Plugins should work Plugins - {Object without require} + options: errors 1`] = `Array []`;

exports[`Options Plugins should work Plugins - {Object without require} + options: warnings 1`] = `Array []`;

exports[`Options Plugins should work Plugins - {Object without require}: css 1`] = `
"a { color: rgba(255, 0, 0, 1.0) }
"
`;

exports[`Options Plugins should work Plugins - {Object without require}: errors 1`] = `Array []`;

exports[`Options Plugins should work Plugins - {Object without require}: warnings 1`] = `Array []`;

exports[`Options Plugins should work Plugins - {Object} + options: css 1`] = `
"a {
  border-top-color: blue;
  border-right-color: blue;
  border-left-color: blue;
  background-color: #fafafa;
}
"
`;

exports[`Options Plugins should work Plugins - {Object} + options: errors 1`] = `Array []`;

exports[`Options Plugins should work Plugins - {Object} + options: warnings 1`] = `Array []`;

exports[`Options Plugins should work Plugins - {Object}: css 1`] = `
"a { color: rgba(255, 0, 0, 1.0) }
"
`;

exports[`Options Plugins should work Plugins - {Object}: errors 1`] = `Array []`;

exports[`Options Plugins should work Plugins - {Object}: warnings 1`] = `Array []`;

exports[`Options Plugins should work Plugins - {empty Object}: css 1`] = `
"a { color: rgba(255, 0, 0, 1.0) }
"
`;

exports[`Options Plugins should work Plugins - {empty Object}: errors 1`] = `Array []`;

exports[`Options Plugins should work Plugins - {empty Object}: warnings 1`] = `Array []`;
